lib/matplotlib/image.py

```diff
@@ -162,203 +210,158 @@ def changed(self):
+            out_height = int(ceil(out_height_base) + 1)
+            extra_width = (out_width - out_width_base) / out_width_base
+            extra_height = (out_height - out_height_base) / out_height_base
+            t += Affine2D().scale(
+                1.0 + extra_width, 1.0 + extra_height)
+        else:
+            out_width = int(out_width_base)
+            out_height = int(out_height_base)
+
+        if not unsampled:
+            if len(A.shape) == 2:
+                A = self.norm(A)
+                if A.dtype.kind == 'f':
+                    # For floating-point greyscale images, we treat negative
+                    # numbers as transparent.
+                    output = np.empty((out_height, out_width), dtype=A.dtype)
+                    output[...] = -100.0
+                else:
+                    output = np.zeros((out_height, out_width), dtype=A.dtype)
 
-        return im, xmin, ymin, dxintv, dyintv, sx, sy
+                alpha = 1.0
+            elif len(A.shape) == 3:
+                # Always convert to RGBA, even if only RGB input
+                if A.shape[2] == 3:
+                    A = _rgb_to_rgba(A)
+                elif A.shape[2] != 4:
+                    raise ValueError("Invalid dimensions, got %s" % (A.shape,))
 
-    @staticmethod
-    def _get_rotate_and_skew_transform(x1, y1, x2, y2, x3, y3):
-        """
-        Retuen a transform that does
-         (x1, y1) -> (x1, y1)
-         (x2, y2) -> (x2, y2)
-         (x2, y1) -> (x3, y3)
-
-        It was intended to derive a skew transform that preserve the
-        lower-left corner (x1, y1) and top-right corner(x2,y2), but
-        change the lower-right-corner(x2, y1) to a new position
-        (x3, y3).
-        """
-        tr1 = mtransforms.Affine2D()
-        tr1.translate(-x1, -y1)
-        x2a, y2a = tr1.transform_point((x2, y2))
-        x3a, y3a = tr1.transform_point((x3, y3))
+                output = np.zeros((out_height, out_width, 4), dtype=A.dtype)
 
-        inv_mat = 1. / (x2a*y3a-y2a*x3a) * np.mat([[y3a, -y2a], [-x3a, x2a]])
+                alpha = self.get_alpha()
+                if alpha is None:
+                    alpha = 1.0
+            else:
+                raise ValueError("Invalid dimensions, got %s" % (A.shape,))
+
+            _image.resample(
+                A, output, t, _interpd_[self.get_interpolation()],
+                self.get_resample(), alpha,
+                self.get_filternorm() or 0.0, self.get_filterrad() or 0.0)
+
+            output = self.to_rgba(output, bytes=True, norm=False)
+
+            # Apply alpha *after* if the input was greyscale
+            if len(A.shape) == 2:
+                alpha = self.get_alpha()
+                if alpha is not None and alpha != 1.0:
+                    alpha_channel = output[:, :, 3]
+                    alpha_channel[:] = np.asarray(
+                        np.asarray(alpha_channel, np.float) * alpha, np.uint8)
```

Comment:
Yes, we're handling alpha correctly here.  We only have premultiplied alpha issues when loading PNG files from disk -- and the unmultiplying happens upon reading.  Everything internally should be unmultiplied by the time we get here.

Good point -- we can probably get away with 32-bit floats here.

`_image.resample` is actually ignorant of alpha on greyscale images (they don't have an alpha channel we can use to store alpha values).  But the negative values are converted to alpha of 0.0 in the subsequent call to `to_rgba`.  Here, we're applying the global alpha value over the whole image (if passed in, which it rarely is).  The values of 0.0 (which were negative in greyscale) will remain 0.0 through the multiplication.